### PR TITLE
Windows: alternate data steams

### PR DESCRIPTION
**Pull request by Anakonda**
Backup and extract Windows alternate data steams (#1342 #983 )
ADS are saved as separate items and extracted as such on non-windows.

### CODE REVIEW

File: src/borg/archiver.py [hunk 1]
```diff
@@ -409,6 +410,13 @@ def _process(self, archive, cache, matcher, exclude_caches, exclude_if_present,
             else:
                 status = '-'  # dry run, item was not backed up
         self.print_file_status(status, path)
+        if sys.platform == 'win32':
+            if ':' not in os.path.basename(path):
+                for stream in get_ads(path):
+                    if stream != '::$DATA':
+                        self._process(archive, cache, matcher, exclude_caches, exclude_if_present,
+                            keep_tag_files, skip_inodes, path + stream[:-6], restrict_dev,
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: what's in the last 6 chars?

File: src/borg/platform/windows.pyx [hunk 2]
```diff
@@ -345,3 +359,21 @@ def sync_dir(path):
...
+    cdef HANDLE searchHandle = FindFirstStreamW(cstrPath, FindStreamInfoStandard, <void*>&data, 0)
+    if searchHandle == <HANDLE>INVALID_HANDLE_VALUE:
+        PyMem_Free(cstrPath)
+        return []
+    ret.append(PyUnicode_FromWideChar(data.cStreamName, -1))
+    while FindNextStreamW(searchHandle, <void*>&data) != 0:
+        ret.append(PyUnicode_FromWideChar(data.cStreamName, -1))
+    errno = GetLastError()
+    if errno != 38:
+        print(errno)
```
Comment (ThomasWaldmann) [inline src/borg/platform/windows.pyx]: print?

File: src/borg/archiver.py (hunk 1 above)
Comment (Anakonda) [inline src/borg/archiver.py]: `:$DATA`

File: src/borg/platform/windows.pyx [hunk 3]
```diff
@@ -345,3 +359,21 @@ def sync_dir(path):
...
+    cdef wchar_t* cstrPath = PyUnicode_AsWideCharString(path, NULL)
+    cdef HANDLE searchHandle = FindFirstStreamW(cstrPath, FindStreamInfoStandard, <void*>&data, 0)
+    if searchHandle == <HANDLE>INVALID_HANDLE_VALUE:
+        PyMem_Free(cstrPath)
+        return []
+    ret.append(PyUnicode_FromWideChar(data.cStreamName, -1))
+    while FindNextStreamW(searchHandle, <void*>&data) != 0:
+        ret.append(PyUnicode_FromWideChar(data.cStreamName, -1))
+    errno = GetLastError()
+    if errno != 38:
```
Comment (enkore) [inline src/borg/platform/windows.pyx]: ERROR_HANDLE_EOF

**Comment by ThomasWaldmann** [comment]
If I would extract a single filename on windows, it would not extract its ADS, right?
So I somehow would need to lookup first whether the file has ADS and then manually extract them all.

**Comment by Anakonda** [comment]
Correct, I don't think that is problem

**Comment by enkore** [comment]
lgtm

**Comment by ThomasWaldmann** [comment]
@Anakonda well, in the case i have given (someone wants to extract a single file), the resulting file does not have the ADS because they are treated like separate files. If one is aware of that, one can workaround this, but if not ...

**Comment by enkore** [comment]
^- I think this is ok (for now?)

Merge?

**Comment by Anakonda** [comment]
I think merge.

**Comment by enkore** [comment]
Thanks, and sorry this took so long.

**Pull request merged**